site/static/api/device-registry-revised.yaml

```diff
@@ -0,0 +1,725 @@
+         additionalProperties: false
+         properties:
+            "enabled":
+               type: boolean
+               default: true
+            "ext":
+               $ref: '#/components/schemas/Extensions'
+            "adapters":
+               type: array
+               description: Only a single entry per type is allowed. If multiple entries for the same type are present it is handled as an error.
+               items:
+                  $ref: '#/components/schemas/Adapter'
+            "limits":
+               $ref: '#/components/schemas/ResourceLimit'
+            "trusted-ca":
+               $ref: '#/components/schemas/TrustedCA'
+
+      TrustedCA:
+         type: object
+         additionalProperties: false
+         required:
+            - "subject-dn"
+         properties:
+            "subject-dn":
+               type: string
+            "public-key":
+               type: string
+               format: byte
+            "cert":
+               type: string
+               format: byte
+
+      Adapter:
+         type: object
+         additionalProperties: true
+         required:
+            - type
+         properties:
+            "type":
+               type: string
+            "enabled":
+               type: boolean
+               default: false
+            "device-authentication-required":
+               type: boolean
+               default: true
+            "ext":
+               $ref: '#/components/schemas/Extensions'
+
+      ResourceLimit:
+         type: object
+         additionalProperties: false
+         properties:
+            "max-connections":
+               type: integer
+               default: -1
+
+# Devices schema
+
+      Device:
+         type: object
+         additionalProperties: false
+         properties:
+            "enabled":
+               type: boolean
+               default: true
+            "ext":
+               $ref: '#/components/schemas/Extensions'
+
+# Credentials
+
+      CredentialsSet:
+         type: object
+         additionalProperties:
+            $ref: '#/components/schemas/AuthenticationIdentity'
+
+      AuthenticationIdentity:
+         type: object
+         required:
+            - secrets
+         properties:
+            "enabled":
+               type: boolean
+               default: true
+            "secrets":
+               type: array
+               items:
+                  $ref: '#/components/schemas/TypedSecret'
+
+      TypedSecret:
+         oneOf:
+            - $ref: '#/components/schemas/PasswordSecret'
+            - $ref: '#/components/schemas/PSKSecret'
+            - $ref: '#/components/schemas/X509CertificateSecret'
+         discriminator:
+            propertyName: type
+            mapping:
+               "hashed-password": '#/components/schemas/PasswordSecret'
+               "psk": '#/components/schemas/PSKSecret'
+               "x509-cert": '#/components/schemas/X509CertificateSecret'
+
+      CommonSecret:
```

Comment:
Maybe I'm missing something, but I think the CommonSecret should extend the TypedSecret. 
E.g.:
```YAML
CommonSecret:
      allOf:
        - $ref: '#/components/schemas/TypedSecret'
        - type: object
          properties:
            "type":
              type: string
            "enabled":
              type: boolean
              default: true
            "not-before":
              type: string
              format: date-time
            "not-after":
              type: string
              format: date-time
            "comment":
              type: string
```
Again, in terms of code generation for java, this would lead to a decoupled TypedSecret and would be hard to get back together with the concrete secrets. Currently you would get something like:

```Java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type", visible = true)
@JsonSubTypes({
  @JsonSubTypes.Type(value = PasswordSecret.class, name = "hashed-password"),
  @JsonSubTypes.Type(value = PSKSecret.class, name = "psk"),
  @JsonSubTypes.Type(value = X509CertificateSecret.class, name = "x509-cert"),
})
public class TypedSecret { ... }

public class CommonSecret { ... }

public class X509CertificateSecret extends CommonSecret { ... }
```
where the "oneOf" information is (in best case) included as annotations, but not in the object hirarchy. The example is for spring(with jackson), if you use the plain java generator (or f.e. the vertx-generator) you get no connection between TypedSecret and Common/PasswordSecret/etc.  at all.

Reply:
I tried to explain this in previous comment. The idea was the following (and I didn't check for the generated code, but then again, code generation always misses something):

Each device has a list of authentication identities, which is defined as a `Map<String,AuthenticationIdentity>`. Each `AuthenticationIdentity` has a list of secrets, which could either be `PasswordSecret`, `PSKSecret`, or some other type.

So the `AuthenticationIdentity` should look like:

~~~java
class AuthenticationIdentity {
  List<TypedSecret> secrets;
}
~~~

The `TypedSecret` is the "union type" for a single secret entry. It is required to have `type`, but none of the other properties from `CommonSecret`.

The other secrets however share the fields from the `CommonSecret`, just because the choose to do so.

So again, looking at Java I would see this as:

~~~java
interface TypedSecret {
   String getType();
   String setType(String type);
}

class CommonSecret {
   public String getType() {
    return type;
  }

  public void setType(String type) {
    this.type = type;
  }
  // ...
}

class PassswordSecret extends CommonSecret {
}
~~~

My experience with the code generators from OpenAPI and Swagger showed that you will always have something that doesn't really work. And if it works fine in language A, it fails in language B. So, I am not sure how far we should push the spec file to produce code. I would see it more as a documentation.

Reply:
But just to be clear. I am open to suggestions on how to improve this!

Reply:
Thanks for the explanation. It seems that this is a shortcoming of the generators, it makes totally sense the way you defined it.

Reply:
@wistefan I am glad to hear that. I did in the meantime play around with the model, based on your proposal and what OpenAPI documents. I ended up with something that worked fine-ish in Java, was messed up both in Go and the actual Specification then (and I didn't test beyond that).

I guess we have to still wait a few centuries, for a modelling approach that just works :wink: 

So for the moment I would focus in the spec itself, and less on the code generation.

Reply:
And just because I didn't say it explicitly. I am thankful for your input! Keep it coming!

Reply:
Especially for java, there are quite some discussions on how to implement the support for one-of(and for that matter all-of and any-of) in the generators. I extended my generator now to produce it as: 
```Java
interface TypedSecret {
   String getType();
   String setType(String type);
}

class CommonSecret {
   public String getType() {
    return type;
  }

  public void setType(String type) {
    this.type = type;
  }
  // ...
}

class PassswordSecret extends CommonSecret implements TypedSecret {
}
```
which seems to work best at the moment. 
I also think it will take a while to for a working modelling approach(if there will be one at all).